kong/status/ready.lua

```diff
@@ -0,0 +1,118 @@
+local declarative = require "kong.db.declarative"
+local constants = require "kong.constants"
+
+local ngx = ngx
+local ngx_log = ngx.log
+local ngx_NOTICE = ngx.NOTICE
+local ngx_DEBUG = ngx.DEBUG
+
+local tonumber = tonumber
+local kong = kong
+local fmt = string.format
+
+local get_current_hash = declarative.get_current_hash
+
+
+local worker_count = ngx.worker.count()
+local kong_shm     = ngx.shared.kong
+
+local is_dbless = kong.configuration.database == "off"
+local is_control_plane = kong.configuration.role == "control_plane"
+
+local PLUGINS_REBUILD_COUNTER_KEY = constants.PLUGINS_REBUILD_COUNTER_KEY
+local ROUTERS_REBUILD_COUNTER_KEY = constants.ROUTERS_REBUILD_COUNTER_KEY
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+
+
+local function is_dbless_ready(router_rebuilds, plugins_iterator_rebuilds)
+  if router_rebuilds < worker_count then
+    return false, fmt("router builds not yet complete, router ready"
+      .. " on %d of %d workers", router_rebuilds, worker_count)
+  end
+
+  if plugins_iterator_rebuilds < worker_count then
+    return false, fmt("plugins iterator builds not yet complete, "
+      .. "plugins iterator ready on %d of %d workers",
+      plugins_iterator_rebuilds, worker_count)
+  end
+
+  local current_hash = get_current_hash()
+
+  if not current_hash then
+    return false, "no configuration available (configuration hash is not initialized)"
+  end
+
+  if current_hash == DECLARATIVE_EMPTY_CONFIG_HASH then
+    return false, "no configuration available (configuration hash is empty)"
+  end
+
+  return true
+end
+
+
+local function is_traditional_ready(router_rebuilds, plugins_iterator_rebuilds)
+    -- data plane with db, only build once, because
+    -- build_router() will not be called for each worker because of ROUTER_CACHE
```

Comment:
```suggestion
    -- traditional mode builds router from database once inside `init` phase
```

Is this more accurate (from my vague memories)?